### PR TITLE
Array bugs2

### PR DESCRIPTION
**Pull request by davek2**
This PR replaces previous PR "fix 2 array bugs", which can be deleted.  This version is based on and integrated with bsokol's updates to master.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,12 @@
-build/*
-dist/*
-genson.egg-info/*
+# See: github.com/github/gitignore/blob/master/Python.gitignore
```
Comment (wolverdude) [inline .gitignore]: :+1: I usually keep this stuff in a global `.gitignore`, but not everyone does, so this is good practice. Thanks for doing the research and compiling this.

File: extras/printschema.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import copy
```
Comment (wolverdude) [inline extras/printschema.py]: This file seems tangential to the rest of the diff. Could you pull it out and put it in a separate PR?

Also, on a side-note. My usual way of pretty-printing here is to use the [`json` npm package](https://www.npmjs.com/package/json):

```
$ echo '{"hi":"there"}' | genson | json -i
{ required: [ 'hi' ],
  type: 'object',
  properties: { hi: { type: 'string' } } }
```

File: genson/generator.py [hunk 3]
```diff
@@ -19,7 +19,7 @@ class Schema(object):
     with existing schemas and objects before being serialized.
     """
 
-    def __init__(self, merge_arrays=True):
+    def __init__(self, merge_arrays=True, _c=None):
```
Comment (wolverdude) [inline genson/generator.py]: Instead of passing a private context argument here, I would just convert everything to `**options`, lazily set some defaults, and then pass it down the chain. That way there's no protected interface that users have to figure out and work around.

File: test/test_gen_single.py [hunk 4]
```diff
@@ -1,181 +1,211 @@
 import unittest
-import os
-import sys
-sys.path.insert(1, os.path.join(sys.path[0], '..'))
-
-from genson import Schema
 import base
 
+def check(self, instance, options, expected):  # alias for readability
```
Comment (wolverdude) [inline test/test_gen_single.py]: I would prefer not to do this. While converting everything to simple, positional arguments takes less space to write, it makes the tests harder to read, and code readability/maintainability is always my # 1 concern.

File: test/base.py [hunk 5]
```diff
@@ -1,12 +1,28 @@
...
-
 import jsonschema
+import os, sys
 
+sys.path.insert(1, os.path.join(sys.path[0], '..'))
+from genson import Schema
 
 class SchemaTestCase(unittest.TestCase):
+
+    def assertGenSchema(self, instance, options, expected):
```
Comment (wolverdude) [inline test/base.py]: Likewise, I would prefer not to have this method. It means writing one or two extra lines in the test, but those lines make it very explicit what the test is doing.

File: test/base.py [hunk 6]
```diff
@@ -1,12 +1,28 @@
...
+        actual = Schema(**options).add_object(instance).to_dict()
+        self.assertSchema(actual, expected)
+        self.assertValidData(instance, actual)
```
Comment (wolverdude) [inline test/base.py]: Since this is part of the user contract, I think I'm okay just putting it inside `assertSchema`.

File: test/base.py [hunk 7]
```diff
@@ -1,12 +1,28 @@
...
+        return actual
+
     def assertSchema(self, actual, expected):
         self.assertValidSchema(actual)
         self.assertEqual(actual, expected)
 
     def assertValidSchema(self, schema):
         jsonschema.Draft4Validator.check_schema(schema)
+
+    def assertValidData(self, data, schema):
```
Comment (wolverdude) [inline test/base.py]: Rename to `assertDictValidatesUnderSchema`, or something similar so that the param order is clear.

File: test/test_gen_single.py [hunk 8]
```diff
@@ -1,181 +1,211 @@
...
-        s = Schema().add_object([])
-        self.assertSchema(s.to_dict(),
-                          {"type": "array"})
+        check(self, [], {}, {"type": "array", "items": {}})
+
+    def test_empty_sep(self):
+        check(self, [], {"merge_arrays": False}, {"type": "array"})
+
+    @unittest.expectedFailure
+    def test_empty_BAD(self):       # Empty array is not a valid schema
```
Comment (wolverdude) [inline test/test_gen_single.py]: Rather than keeping bad tests around, just remove them.

**Comment by wolverdude** [comment]
Thanks for submitting this diff, and thanks again for discovering the bugs! I did a once-over on it and made some comments. Haven't had a chance to look over the test cases yet. Might do that later.

File: test/base.py (hunk 6 above)
Comment (davek2) [inline test/base.py]: Because test_gen_multi calls assertSchema, I don't believe the call signature can be changed.  The single-step "generate a schema from a single object, validate the schema, and validate the object" call is useful but doesn't supersede the standalone "validate the schema" call.

Comment (wolverdude) [inline test/base.py]: Ah, yes. We don't have access to the base object in that case. Yeah, this will have to stay separate. It's a good thing to test for though, and I see now why you created this method. There should be a way that we can abstract this into a method that we can also use it in `TestGenMulti` without sacrificing clarity. Let's think on it.

File: test/test_gen_single.py (hunk 8 above)
Comment (davek2) [inline test/test_gen_single.py]: Deleted 2 expected failure tests.

File: test/test_gen_single.py (hunk 4 above)
Comment (davek2) [inline test/test_gen_single.py]: Done.

File: extras/printschema.py (hunk 2 above)
Comment (davek2) [inline extras/printschema.py]: Hmm, I'm a git newbie.  I could just delete the file from HEAD, but that wouldn't make commit 62d2b30 disappear.  Is that OK or did you want something fancier?

I'm using genson to analyze large (tens of KBs, some up to 1MB) data objects and using the generated schemas as documentation.  With the python module I can put all schemas in a consistent format -- type first, then required, then properties,  required and basic types on one line, nested types indented on multiple lines, etc).  The schema-aware pprinter will handle any dict, but it does a better job than generic pprinters on dicts that are schemas.  You don't need to include it with genson; I just thought someone might be interested.

Comment (wolverdude) [inline extras/printschema.py]: wrt Git: This file is small, so it doesn't bother me that it'll stay in git history. If you want to remove it from history though, you can use a rebase like [this](http://sethrobertson.github.io/GitFixUm/fixup.html#remove_deep). It can get kinda messy though, especially since you also modified the `.gitignore`. Plus you'd need to force-push, which would mess up this diff. IMHO, all this trouble isn't worth it.

Comment (wolverdude) [inline extras/printschema.py]: wrt the idea: Thanks for explaining your use-case to me. It's great to hear how people are using this library!

I don't know that this belongs in the library itself though, but I'd love to support that functionality in a flexible way where you don't have to literally hack the library to make it work. Maybe we could add a hook to enable custom renderers, something like the way it works in the [`json` library](https://docs.python.org/3/library/json.html#json.JSONEncoder), but hopefully a little simpler.

Still I don't anticipate this making version `0.2`. I really want to get those bugfixes out there ASAP, so we'll wait until the next version for this.

File: genson/generator.py (hunk 3 above)
Comment (davek2) [inline genson/generator.py]: Thanks for the suggestion - that's SO much cleaner.

Comment (davek2) [inline genson/generator.py]: Thanks for the suggestion - that is SO much cleaner.

**Comment by wolverdude** [comment]
Just FYI: I haven't forgotten about this. I just got busy with work, Easter, Taxes, etc. I will get to it, but it may be a couple more weeks before I do.

File: genson/generator.py [hunk 9]
```diff
@@ -193,28 +197,26 @@ def _add_properties(self, properties, func):
...
     def _add_items_sep(self, items, func):
         for item in items:
-            subschema = Schema()
+            subschema = Schema(**self._kwargs)
             getattr(subschema, func)(item)
-
-            # only add schema if it's not already there.
-            if subschema not in self._items:
-                self._items.append(subschema)
+            self._items.append(subschema)
```
Comment (wolverdude) [inline genson/generator.py]: This doesn't actually merge two objects' schemas in a way that's compatible with both. Only the first one will be fully served.

File: test/base.py (hunk 5 above)
Comment (wolverdude) [inline test/base.py]: I figured out a way to simplify the testing interface using test instance state. I'll add it after the merge.

**Comment by wolverdude** [comment]
I'll go ahead and merge this. Then I'll fix the two issues I brought up.
